File: Polizing-Dashbord/scripts/test-tarjeta-felder.ts
/**
 * Script de prueba: pide la tarjeta de circulación para FELDER DAVID.
 * Uso: npx tsx scripts/test-tarjeta-felder.ts
 */

import "dotenv/config";
import { Pool } from "pg";
import { PrismaPg } from "@prisma/adapter-pg";
import { PrismaClient } from "../app/generated/prisma/client";
import { normalizePgUrl } from "../lib/utils/db-url";

const API_URL = "https://polizing-interno.vercel.app";
const API_KEY = process.env.CHATBOT_API_KEY ?? "";

async function main() {
  const url = process.env.DIRECT_URL ?? process.env.DATABASE_URL;
  if (!url) throw new Error("Missing DIRECT_URL/DATABASE_URL");

  const pool = new Pool({ connectionString: normalizePgUrl(url) });
  const prisma = new PrismaClient({ adapter: new PrismaPg(pool) });

  try {
    // 1. Buscar FELDER DAVID por apellido + nombre
    const nc = await prisma.clientes_no_corporativos.findFirst({
      where: {
        OR: [
          {
            apellido: { contains: "FELDER", mode: "insensitive" },
            nombre: { contains: "DAVID", mode: "insensitive" },
          },
          {
            apellido: { contains: "DAVID", mode: "insensitive" },
            nombre: { contains: "FELDER", mode: "insensitive" },
          },
        ],
      },
      include: { cliente: true },
    });

    if (!nc) {
      console.error("❌ No se encontró FELDER DAVID en clientes_no_corporativos");
      return;
    }

    const cliente = nc.cliente;
    console.log(`✅ Cliente encontrado: id=${cliente.id}  teléfono=${cliente.telefono}  estado=${cliente.estado}`);
    console.log(`   Nombre: ${nc.nombre} ${nc.apellido}`);

    if (!cliente.telefono) {
      console.error("❌ El cliente no tiene teléfono cargado");
      return;
    }

    // 2. Listar pólizas vigentes
    const polizas = await prisma.polizas.findMany({
      where: {
        cliente_id: cliente.id,
        estado: { in: ["vigente", "proxima"] },
      },
      select: {
        id: true,
        numero_poliza: true,
        estado: true,
        rama: true,
        suplemento: true,
        aseguradora_id: true,
        tarjeta_circulacion_pdf: true,
      },
      orderBy: { id: "asc" },
    });

    if (polizas.length === 0) {
      console.error("❌ No hay pólizas vigentes/próximas para este cliente");
      return;
    }

    console.log(`\n📋 Pólizas vigentes (${polizas.length}):`);
    for (const p of polizas) {
      const hasPdf = p.tarjeta_circulacion_pdf ? "✅ PDF cacheado" : "⚠️  sin PDF";
      console.log(`   [${p.id}] ${p.numero_poliza}  estado=${p.estado}  rama=${p.rama}  ${hasPdf}`);
    }

    // 3. Pedir tarjeta para cada póliza
    console.log("\n🔄 Solicitando tarjetas al endpoint...\n");
    for (const poliza of polizas) {
      console.log(`→ Póliza ${poliza.numero_poliza} (id=${poliza.id})`);
      const res = await fetch(`${API_URL}/api/chatbot/circulation-card`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-API-Key": API_KEY,
        },
        body: JSON.stringify({ phone: cliente.telefono, policy_id: poliza.id }),
      });

      const body = await res.json().catch(() => null);

      if (!res.ok) {
        console.log(`  ❌ HTTP ${res.status}: ${JSON.stringify(body)}`);
        continue;
      }

      if (body?.mode === "link") {
        console.log(`  ✅ mode=link`);
        console.log(`     source_url: ${body.source_url}`);
        console.log(`     filename:   ${body.filename}`);
      } else if (body?.mode === "document") {
        const b64Preview = (body.content_base64 as string)?.slice(0, 60) ?? "";
        console.log(`  ✅ mode=document (PDF cacheado)`);
        console.log(`     filename:   ${body.filename}`);
        console.log(`     mime:       ${body.mime_type}`);
        console.log(`     base64:     ${b64Preview}...`);
      } else {
        console.log(`  ⚠️  Respuesta inesperada:`, JSON.stringify(body, null, 2));
      }

      console.log();
    }
  } finally {
    await pool.end();
  }
}

main().catch((err) => {
  console.error("Error fatal:", err);
  process.exit(1);
});

File: Polizing-Dashbord/scripts/test_tarjeta_felder.py
"""
Test: pide la tarjeta de circulación para FELDER DAVID.

Uso (desde D:/Polizing-Interno/Polizing-Dashbord):
    python scripts/test_tarjeta_felder.py
"""

import os
import re
import json
from urllib.parse import urlparse, unquote

import httpx
import psycopg2
import psycopg2.extras


# ── Config ──────────────────────────────────────────────────────────────────

def load_env(path=".env"):
    """Carga variables de un archivo .env en os.environ."""
    try:
        with open(path, encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("#"):
                    continue
                if "=" in line:
                    k, _, v = line.partition("=")
                    v = v.strip().strip('"').strip("'")
                    os.environ.setdefault(k.strip(), v)
    except FileNotFoundError:
        pass


def parse_pg_url(url: str):
    """Parsea un connection string de Postgres con password que puede contener @."""
    proto_end = url.index("://") + 3
    rest = url[proto_end:]
    last_at = rest.rfind("@")
    user_info = rest[:last_at]
    host_part = rest[last_at + 1:]

    colon = user_info.index(":")
    user = user_info[:colon]
    password = unquote(user_info[colon + 1:])

    # host_part: host:port/dbname?params
    m = re.match(r"([^:/]+):(\d+)/([^?]+)", host_part)
    if not m:
        raise ValueError(f"No se pudo parsear host de: {host_part}")
    host, port, dbname = m.group(1), int(m.group(2)), m.group(3)
    return dict(host=host, port=port, dbname=dbname, user=user, password=password)


def get_db():
    url = os.environ.get("DIRECT_URL") or os.environ.get("DATABASE_URL")
    if not url:
        raise RuntimeError("Falta DIRECT_URL / DATABASE_URL")
    return psycopg2.connect(**parse_pg_url(url))


API_BASE = "https://polizing-interno.vercel.app"


# ── Main ─────────────────────────────────────────────────────────────────────

def main():
    load_env()
    api_key = os.environ.get("CHATBOT_API_KEY", "")

    conn = get_db()
    cur = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)

    # 1. Buscar FELDER DAVID
    cur.execute("""
        SELECT c.id, c.telefono, c.estado,
               nc.nombre, nc.apellido
        FROM chatbot.clientes c
        JOIN chatbot.clientes_no_corporativos nc ON nc.cliente_id = c.id
        WHERE (UPPER(nc.apellido) LIKE '%FELDER%' AND UPPER(nc.nombre) LIKE '%DAVID%')
           OR (UPPER(nc.apellido) LIKE '%DAVID%'  AND UPPER(nc.nombre) LIKE '%FELDER%')
        ORDER BY c.id ASC
        LIMIT 5;
    """)
    rows = cur.fetchall()

    if not rows:
        print("❌ No se encontró FELDER DAVID en clientes_no_corporativos")
        cur.close(); conn.close()
        return

    for row in rows:
        cliente_id = row["id"]
        telefono = row["telefono"]
        print(f"✅ Cliente encontrado: id={cliente_id}  nombre={row['nombre']} {row['apellido']}")
        print(f"   teléfono={telefono}  estado={row['estado']}")

        if not telefono:
            print("   ❌ Sin teléfono — saltando\n")
            continue

        # 2. Pólizas vigentes
        cur.execute("""
            SELECT id, numero_poliza, estado, rama, suplemento, aseguradora_id,
                   (tarjeta_circulacion_pdf IS NOT NULL) as has_pdf
            FROM chatbot.polizas
            WHERE cliente_id = %s AND estado IN ('vigente', 'proxima')
            ORDER BY id ASC;
        """, (cliente_id,))
        polizas = cur.fetchall()

        if not polizas:
            print("   ❌ Sin pólizas vigentes/próximas\n")
            continue

        print(f"\n   📋 Pólizas vigentes ({len(polizas)}):")
        for p in polizas:
            pdf_tag = "✅ PDF" if p["has_pdf"] else "⚠️  sin PDF"
            print(f"     [{p['id']}] {p['numero_poliza']}  estado={p['estado']}  rama={p['rama']}  {pdf_tag}")

        # 3. Llamar al endpoint por cada póliza
        print("\n   🔄 Solicitando tarjetas...\n")
        headers = {"Content-Type": "application/json", "X-API-Key": api_key}

        for p in polizas:
            payload = {"phone": telefono, "policy_id": p["id"]}
            print(f"   → Póliza {p['numero_poliza']} (id={p['id']})")

            r = httpx.post(
                f"{API_BASE}/api/chatbot/circulation-card",
                headers=headers,
                json=payload,
                timeout=30,
            )

            try:
                body = r.json()
            except Exception:
                body = None

            if not r.is_success:
                print(f"     ❌ HTTP {r.status_code}: {body}\n")
                continue

            mode = body.get("mode") if body else None
            if mode == "link":
                print(f"     ✅ mode=link")
                print(f"        source_url: {body['source_url']}")
                print(f"        filename:   {body['filename']}")
            elif mode == "document":
                b64 = (body.get("content_base64") or "")[:60]
                print(f"     ✅ mode=document (PDF cacheado)")
                print(f"        filename:   {body['filename']}")
                print(f"        mime:       {body['mime_type']}")
                print(f"        base64[:60]: {b64}...")
            else:
                print(f"     ⚠️  Respuesta inesperada: {json.dumps(body, indent=6)}")
            print()

        print()

    cur.close()
    conn.close()


if __name__ == "__main__":
    main()
